Reduce call count of `<=>` in Array#minmax

mrkn opened the pull request:
This patch can reduce the call count of `<=>` in `Array#minmax`.

The benchmark result is below:

```
# Iteration per second (i/s)
|                 |compare-ruby|built-ruby|
|:----------------|-----------:|---------:|
|ary2.minmax      |      4.175M|    5.333M|
|                 |           -|     1.28x|
|ary10.minmax     |      3.464M|    4.274M|
|                 |           -|     1.23x|
|ary100.minmax    |      1.187M|    2.008M|
|                 |           -|     1.69x|
|ary500.minmax    |    296.217k|  538.213k|
|                 |           -|     1.82x|
|ary1000.minmax   |    154.837k|  268.651k|
|                 |           -|     1.74x|
|ary2000.minmax   |     68.807k|   83.958k|
|                 |           -|     1.22x|
|ary3000.minmax   |     64.350k|   85.220k|
|                 |           -|     1.32x|
|ary5000.minmax   |     34.222k|   36.934k|
|                 |           -|     1.08x|
|ary10000.minmax  |     17.458k|   16.978k|
|                 |       1.03x|         -|
|ary20000.minmax  |      8.743k|    8.217k|
|                 |       1.06x|         -|
|ary50000.minmax  |      3.287k|    3.164k|
|                 |       1.04x|         -|
```

I don't know why this change doesn't affect an array which has over 10000 items.

mrkn commented:
The latest benchmark result is below:

```
# Iteration per second (i/s)

|                 |compare-ruby|built-ruby|
|:----------------|-----------:|---------:|
|ary2.minmax      |      3.210M|    4.279M|
|                 |           -|     1.33x|
|ary10.minmax     |      2.417M|    3.524M|
|                 |           -|     1.46x|
|ary100.minmax    |    970.214k|    2.116M|
|                 |           -|     2.18x|
|ary500.minmax    |    315.657k|  545.792k|
|                 |           -|     1.73x|
|ary1000.minmax   |    163.599k|  277.485k|
|                 |           -|     1.70x|
|ary2000.minmax   |     68.385k|   85.358k|
|                 |           -|     1.25x|
|ary3000.minmax   |     69.948k|   83.187k|
|                 |           -|     1.19x|
|ary5000.minmax   |     31.845k|   37.180k|
|                 |           -|     1.17x|
|ary10000.minmax  |     17.245k|   18.326k|
|                 |           -|     1.06x|
|ary20000.minmax  |      8.659k|    8.825k|
|                 |           -|     1.02x|
|ary50000.minmax  |      3.430k|    3.547k|
|                 |           -|     1.03x|
```

The above result uses clang to build ruby.  I'll check the result also by gcc later.

mrkn commented:
I investigated the benchmark result by gcc, and made the following figure to show the relationship between the array size and the ratio of IPS improvements.

![bench](https://user-images.githubusercontent.com/3959/85245609-17b9f600-b483-11ea-94cd-15b55d936932.png)

mrkn commented:
@shyouhei Oh, I didn't notice that the results of benchmark-driver can be interpreted as tables. Thank you!

mame commented:
This is interesting, but unfortunately, this change makes `Array#minmax` slower than master when the length of the array is long maybe due to branch-misses.

```
$ perf stat ./miniruby.master -e 'srand(0); a = [*1..1000000].shuffle.map {|n| n.to_s }; 100.times { a.minmax }'
 Performance counter stats for './miniruby.master -e srand(0); a = [*1..1000000].shuffle.map {|n| n.to_s }; 100.times { a.minmax }':
       3513.416512      task-clock (msec)         #    1.000 CPUs utilized
                 3      context-switches          #    0.001 K/sec
                 0      cpu-migrations            #    0.000 K/sec
             21202      page-faults               #    0.006 M/sec
       10401200216      cycles                    #    2.960 GHz
       26345183585      instructions              #    2.53  insn per cycle
        9247699638      branches                  # 2632.110 M/sec
           7916102      branch-misses             #    0.09% of all branches
       3.513818053 seconds time elapsed

$ perf stat ./miniruby.3246 -e 'srand(0); a = [*1..1000000].shuffle.map {|n| n.to_s }; 100.times { a.minmax }'
 Performance counter stats for './miniruby.3246 -e srand(0); a = [*1..1000000].shuffle.map {|n| n.to_s }; 100.times { a.minmax }':
       3649.543797      task-clock (msec)         #    1.000 CPUs utilized
                 4      context-switches          #    0.001 K/sec
                 0      cpu-migrations            #    0.000 K/sec
             21202      page-faults               #    0.006 M/sec
       10704511770      cycles                    #    2.933 GHz
       21473661479      instructions              #    2.01  insn per cycle
        7274559133      branches                  # 1993.279 M/sec
          32561467      branch-misses             #    0.45% of all branches
       3.649905647 seconds time elapsed
```

We need to tweak the code more and carefully evaluate the performance.

mrkn commented:
This is the lates result.

|                   |compare-ruby|built-ruby|
|:------------------|-----------:|---------:|
|ary2.minmax        |      1.791M|    1.803M|
|                   |           -|     1.01x|
|ary10.minmax       |      1.805M|    1.960M|
|                   |           -|     1.09x|
|ary100.minmax      |    882.223k|    1.576M|
|                   |           -|     1.79x|
|ary500.minmax      |    305.951k|  864.080k|
|                   |           -|     2.82x|
|ary1000.minmax     |    148.315k|  607.644k|
|                   |           -|     4.10x|
|ary2000.minmax     |     71.154k|  271.415k|
|                   |           -|     3.81x|
|ary3000.minmax     |     71.254k|  267.795k|
|                   |           -|     3.76x|
|ary5000.minmax     |     33.774k|  114.424k|
|                   |           -|     3.39x|
|ary10000.minmax    |     17.928k|   37.145k|
|                   |           -|     2.07x|
|ary20000.minmax    |      9.075k|   15.948k|
|                   |           -|     1.76x|
|ary50000.minmax    |      3.634k|    6.274k|
|                   |           -|     1.73x|
|ary100000.minmax   |      1.736k|    3.072k|
|                   |           -|     1.77x|
|ary1000000.minmax  |     173.643|   298.495|
|                   |           -|     1.72x|

mrkn commented:
I added the optimization and benchmark cases of `Array#min` and `Array#max`.  Both methods are constantly more than 2 times faster than the original for long arrays.

Moreover, I added benchmark cases of `[ary.min, ary.max]` for comparing with `Array#minmax`.  The result is below. Later I will add the swarmplot of multiple results like above.

|                     |compare-ruby|built-ruby|
|:--------------------|-----------:|---------:|
|ary2.minmax          |      1.936M|    2.087M|
|                     |           -|     1.08x|
|ary10.minmax         |      1.888M|    1.743M|
|                     |       1.08x|         -|
|ary100.minmax        |    944.822k|    1.195M|
|                     |           -|     1.27x|
|ary500.minmax        |    279.681k|  808.211k|
|                     |           -|     2.89x|
|ary1000.minmax       |    155.082k|  615.764k|
|                     |           -|     3.97x|
|ary2000.minmax       |     70.998k|  294.959k|
|                     |           -|     4.15x|
|ary3000.minmax       |     68.769k|  294.421k|
|                     |           -|     4.28x|
|ary5000.minmax       |     36.051k|  112.567k|
|                     |           -|     3.12x|
|ary10000.minmax      |     18.215k|   36.613k|
|                     |           -|     2.01x|
|ary20000.minmax      |      8.964k|   15.966k|
|                     |           -|     1.78x|
|ary50000.minmax      |      3.606k|    5.847k|
|                     |           -|     1.62x|
|ary100000.minmax     |      1.494k|    2.819k|
|                     |           -|     1.89x|
|ary1000000.minmax    |     171.228|   276.443|
|                     |           -|     1.61x|
|min_max(ary2)        |      2.587M|    2.341M|
|                     |       1.11x|         -|
|min_max(ary10)       |      1.310M|    1.576M|
|                     |           -|     1.20x|
|min_max(ary100)      |    862.813k|    1.538M|
|                     |           -|     1.78x|
|min_max(ary1000)     |    161.721k|  356.037k|
|                     |           -|     2.20x|
|min_max(ary2000)     |     70.920k|  166.359k|
|                     |           -|     2.35x|
|min_max(ary3000)     |     63.991k|  157.408k|
|                     |           -|     2.46x|
|min_max(ary5000)     |     34.622k|   88.437k|
|                     |           -|     2.55x|
|min_max(ary10000)    |     17.480k|   44.531k|
|                     |           -|     2.55x|
|min_max(ary100000)   |      1.672k|    4.364k|
|                     |           -|     2.61x|
|min_max(ary1000000)  |     169.064|   400.794|
|                     |           -|     2.37x|
